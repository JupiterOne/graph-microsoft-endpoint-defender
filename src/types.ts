import {
  IntegrationInstanceConfig,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The Azure application client ID used to identify the program as the
   * registered app. This will be the same for all tenants since this is a
   * multi-tenant application.
   */
  clientId: string;

  /**
   * The Azuser application client secret used to authenticate requests.
   */
  clientSecret: string;

  /**
   * The target directory/tenant ID, identified during the admin consent OAuth
   * flow.
   */
  tenant: string;

  /**
   * This is to differentiate between the azure api vs defender api **/
  isDefenderApi: boolean;
}

/**
 * An `IntegrationStepExecutionContext` typed for this integration's
 * `IntegrationInstanceConfig`.
 */
export type IntegrationStepContext =
  IntegrationStepExecutionContext<IntegrationConfig>;

export interface Machine {
  id: string;
  computerDnsName: string;
  firstSeen: string;
  lastSeen: string;
  classification: string;
  onboardingStatus: string;
  healthStatus: string;
  deviceValue: string;
  agentVersion: string;
  defenderAvStatus: string;
  riskScore: string;
  osPlatform: string;
  osVersion: any;
  osProcessor: string;
  version: string;
  lastIpAddress: string;
  lastExternalIpAddress: string;
  osBuild: number;
  rbacGroupId: number;
  rbacGroupName: string;
  exposureLevel: string;
  isAadJoined: boolean;
  aadDeviceId: any;
  machineTags: Array<string>;
  osArchitecture: string;
  managedBy: string;
  managedByStatus: string;
  ipAddresses: Array<IpAddress>;
  vmMetadata: VmMetadata;
}

export interface IpAddress {
  ipAddress: string;
  macAddress: string;
  type: string;
  operationalStatus: string;
}

export interface VmMetadata {
  vmId: string;
  cloudProvider: string;
  resourceId: string;
  subscriptionId: string;
}

export interface UserLogon {
  id: string;
  accountName: string;
  accountDomain: string;
  firstSeen: string;
  lastSeen: string;
  logonTypes: string;
  healthStatus: string;
  deviceValue: string;
  agentVersion: string;
  defenderAvStatus: string;
  isDomainAdmin: boolean;
  logOnMachinesCount: any;
  isOnlyNetworkUser: any;
  accountSid: any;
}

export interface Vulnerability {
  id: any;
  name: string;
  description: string;
  severity: string;
  cvssV3: any;
  exposedMachines: string;
  publishedOn: any;
  updatedOn: any;
  publicExploit: boolean;
  exploitVerified: boolean;
  exploitInKit: false;
  exploitTypes: Array<string>;
  exploitUris: Array<string>;
}

export interface Finding {
  id: any;
  name: string;
  description: string;
  severity: string;
  cvssV3: any;
  exposedMachines: string;
  publishedOn: any;
  updatedOn: any;
  publicExploit: boolean;
  exploitVerified: boolean;
  exploitInKit: false;
  exploitTypes: Array<string>;
  exploitUris: Array<string>;
}
