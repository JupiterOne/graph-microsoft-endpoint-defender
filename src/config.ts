import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationStepExecutionContext,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { GraphClient } from './ms-graph/client';
import { IntegrationInstanceConfigFieldMap } from '@jupiterone/integration-sdk-core';

/**
 * A type describing the configuration fields required to execute the
 * integration for a specific account in the data provider.
 *
 * When executing the integration in a development environment, these values may
 * be provided in a `.env` file with environment variables. For example:
 *
 * - `CLIENT_ID=123` becomes `instance.config.clientId = '123'`
 * - `CLIENT_SECRET=abc` becomes `instance.config.clientSecret = 'abc'`
 *
 * Environment variables are NOT used when the integration is executing in a
 * managed environment. For example, in JupiterOne, users configure
 * `instance.config` in a UI.
 */

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  clientId: {
    type: 'string',
  },
  clientSecret: {
    type: 'string',
    mask: true,
  },
  tenant: {
    type: 'string',
  },
  findingSeverity: {
    type: 'string',
  },
  findingsLimit: {
    type: 'string',
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The Azure application client ID used to identify the program as the
   * registered app. This will be the same for all tenants since this is a
   * multi-tenant application.
   */
  clientId: string;

  /**
   * The Azuser App Registration client secret used to authenticate requests.
   */
  clientSecret: string;

  /**
   * The target directory/tenant ID, identified during the admin consent OAuth
   * flow.
   */
  tenant: string;
}

/**
 * An `IntegrationStepExecutionContext` typed for this integration's
 * `IntegrationInstanceConfig`.
 */
export type IntegrationStepContext =
  IntegrationStepExecutionContext<IntegrationConfig>;

export function validateExecutionConfig(
  executionContext: IntegrationExecutionContext<IntegrationConfig>,
): void {
  const { clientId, clientSecret, tenant } = executionContext.instance.config;

  executionContext.logger.info(
    {
      clientId,
      tenantId: tenant,
    },
    'Configured to make Microsoft Graph API calls to tenantId acting as clientId',
  );

  if (!clientId || !clientSecret || !tenant) {
    throw new IntegrationValidationError(
      'Config requires all of {clientId, clientSecret, tenant}',
    );
  }
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  validateExecutionConfig(context);
  const apiClient = new GraphClient(context.logger, context.instance.config);
  await apiClient.verifyAuthentication();
}
