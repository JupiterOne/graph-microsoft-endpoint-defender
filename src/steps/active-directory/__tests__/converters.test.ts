import { Machine, Finding } from '../../../types';
import {
  createAccountEntityWithOrganization,
  createAccountEntity,
  createMachineEntity,
  createMachinesDeviceRelationship,
  generateUserKey,
  createFindingEntity,
  createMachineUserRelationship,
  createAccountMachineRelationship,
  createFindingsCveRelationship,
  createMachineFindingsRelationship,
} from '../converters';
import { Organization, User } from '@microsoft/microsoft-graph-types';
import { Entity, IntegrationInstance } from '@jupiterone/integration-sdk-core';
import { integrationConfig } from '../../../../test/config';
const exampleOrganization: Organization = {};
const instance: IntegrationInstance = {
  id: '',
  accountId: '',
  name: '',
  integrationDefinitionId: '',
  config: integrationConfig,
};
const exampleAccountEntity: Entity = {
  _class: 'Account',
  _key: 'microsoft_defender_account_id',
  _type: 'microsoft_defender_account',
  displayName: 'name',
};
describe('createAccountEntityWithOrganization', () => {
  test('properties transferred', () => {
    const exampleIntuneConfig = {
      mobileDeviceManagementAuthority: 'mobileDeviceManagementAuthority',
      subscriptionState: 'subscriptionState',
      intuneAccountID: 'intuneAccountID',
    };
    const accountEntity = createAccountEntityWithOrganization(
      instance,
      exampleOrganization,
      exampleIntuneConfig,
    );
    expect(accountEntity).toMatchSnapshot(
      'createAccountEntityWithOrganization',
    );
  });
});

describe('createAccountEntity', () => {
  test('properties transferred', () => {
    expect(createAccountEntity(instance)).toEqual({
      _class: ['Account'],
      _key: 'microsoft_defender_account-',
      _type: 'microsoft_defender_account',
      createdOn: undefined,
      displayName: '',
      name: '',
      _rawData: [],
    });
  });
});

describe('createMachineEntity', () => {
  test('properties transferred', () => {
    const VmMetadata = {
      vmId: 'string',
      cloudProvider: 'string',
      resourceId: 'string',
      subscriptionId: 'string',
    };
    const rawData: Machine = {
      id: 'string',
      computerDnsName: 'string',
      firstSeen: 'string',
      lastSeen: 'string',
      classification: 'string',
      onboardingStatus: 'string',
      healthStatus: 'string',
      deviceValue: 'string',
      agentVersion: 'string',
      defenderAvStatus: 'string',
      riskScore: 'string',
      osPlatform: 'string',
      osVersion: 'any',
      osProcessor: 'string',
      version: 'string',
      lastIpAddress: 'string',
      lastExternalIpAddress: 'string',
      osBuild: 1234,
      rbacGroupId: 456,
      rbacGroupName: 'string',
      exposureLevel: 'string',
      isAadJoined: true,
      aadDeviceId: 'any',
      machineTags: [],
      osArchitecture: 'string',
      managedBy: 'string',
      managedByStatus: 'string',
      ipAddresses: [],
      vmMetadata: VmMetadata,
    };
    expect(createMachineEntity(rawData)).toEqual({
      _class: ['HostAgent'],
      _key: 'string',
      _rawData: [],
      _type: 'microsoft_defender_machine',
      agentVersion: 'string',
      computerDnsName: 'string',
      createdOn: undefined,
      defenderAvStatus: 'string',
      displayName: 'string',
      function: ['endpoint-protection', 'vulnerability-detection'],
      id: 'string',
      name: 'string',
      riskScore: 'string',
    });
  });
});

describe('createMachinesDeviceRelationship', () => {
  test('properties transferred', () => {
    expect(createMachinesDeviceRelationship(exampleAccountEntity)).toEqual({
      _class: 'MANAGES',
      _key: 'microsoft_defender_account_id|manages|FORWARD:_type=user_endPoint:_class=Device,Host:hostname=undefined',
      _mapping: {
        relationshipDirection: 'FORWARD',
        skipTargetCreation: true,
        sourceEntityKey: 'microsoft_defender_account_id',
        targetEntity: {
          _class: ['Device', 'Host'],
          _type: 'user_endPoint',
          category: 'endpoint',
          displayName: undefined,
          fqdn: undefined,
          hostname: undefined,
          id: undefined,
          name: 'name',
          osName: undefined,
          osVersion: undefined,
          platform: undefined,
        },
        targetFilterKeys: [['_type', '_class', 'hostname']],
      },
      _type: 'microsoft_defender_machine_manages_user_endPoint',
      displayName: 'MANAGES',
    });
  });
});
const userMock: User = {};
describe('generateUserKey', () => {
  test('properties transferred', () => {
    expect(generateUserKey(userMock)).toEqual(undefined);
  });
});
const findingMock: Finding = {
  id: '12345',
  name: 'abc',
  description: 'abc',
  severity: 'high',
  cvssV3: '123acdf',
  exposedMachines: 'none',
  publishedOn: '01/02/12',
  updatedOn: '02/05/22',
  publicExploit: true,
  exploitVerified: false,
  exploitInKit: false,
  exploitTypes: [],
  exploitUris: [],
};
describe('createFindingEntity', () => {
  test('properties transferred', () => {
    expect(createFindingEntity(findingMock)).toBeTruthy();
  });
});
const machine: Entity = {
  _class: 'Machine',
  _key: 'microsoft_defender_machine_id',
  _type: 'microsoft_defender_machine',
  displayName: 'name',
};
const user: Entity = {
  _class: 'User',
  _key: 'microsoft_defender_user_id',
  _type: 'microsoft_defender_user',
  displayName: 'name',
};

const finding: Entity = {
  _class: 'Finding',
  _key: 'microsoft_defender_finding_id',
  _type: 'microsoft_defender_finding',
  displayName: 'name',
};

const account: Entity = {
  _class: 'Account',
  _key: 'microsoft_defender_account_id',
  _type: 'microsoft_defender_account',
  displayName: 'name',
};

describe('createMachineUserRelationship', () => {
  test('properties transferred for users', () => {
    expect(createMachineUserRelationship(machine, user)).toEqual({
      _fromEntityKey: 'microsoft_defender_machine_id',
      _key: 'microsoft_defender_machine_id|has|microsoft_defender_user_id',
      _toEntityKey: 'microsoft_defender_user_id',
      _type: 'microsoft_defender_machine_has_user',
      displayName: 'HAS',
      _class: 'HAS',
    });
  });
});

describe('createAccountMachineRelationship', () => {
  test('properties transferred for users', () => {
    expect(createAccountMachineRelationship(machine, account)).toEqual({
      _fromEntityKey: 'microsoft_defender_machine_id',
      _key: 'microsoft_defender_machine_id|has|microsoft_defender_account_id',
      _toEntityKey: 'microsoft_defender_account_id',
      _type: 'microsoft_defender_machine_has_account',
      displayName: 'HAS',
      _class: 'HAS',
    });
  });
});

describe('createFindingsCveRelationship', () => {
  test('properties transferred for users', () => {
    expect(createFindingsCveRelationship(machine)).toEqual({
      _class: 'IS',
      _key: 'microsoft_defender_machine_id|is|cve_microsoft_defender_machine_id',
      _mapping: {
        relationshipDirection: 'FORWARD',
        skipTargetCreation: true,
        sourceEntityKey: 'microsoft_defender_machine_id',
        targetEntity: {
          Resource: 'CVE',
          _class: ['Vulnerability'],
          _key: 'cve_microsoft_defender_machine_id',
          _type: 'cve',
          displayName: 'name',
        },
        targetFilterKeys: [['Resource', '_type', '_class', '_key']],
      },
      _type: 'microsoft_defender_vulnerability_is_cve',
      displayName: 'IS',
    });
  });
});

describe('createMachineFindingsRelationship', () => {
  test('properties transferred for users', () => {
    expect(createMachineFindingsRelationship(machine, finding)).toEqual({
      _key: 'microsoft_defender_machine_id|identified|microsoft_defender_finding_id',
      _type: 'microsoft_defender_machine_identified_finding',
      displayName: 'IDENTIFIED',
      _class: 'IDENTIFIED',
      _fromEntityKey: 'microsoft_defender_machine_id',
      _toEntityKey: 'microsoft_defender_finding_id',
    });
  });
});
