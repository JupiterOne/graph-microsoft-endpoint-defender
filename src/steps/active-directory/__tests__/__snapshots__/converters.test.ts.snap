// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAccountEntityWithOrganization properties transferred: createAccountEntityWithOrganization 1`] = `
Object {
  "_class": Array [
    "Account",
  ],
  "_key": "microsoft_defender_account-5a721b05-53ed-4ed9-be02-aed28f11edbd",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "organization": Object {},
      },
    },
  ],
  "_type": "microsoft_defender_account",
  "createdOn": undefined,
  "defaultDomain": undefined,
  "displayName": "DefenderAccount",
  "id": undefined,
  "name": undefined,
  "organizationName": undefined,
  "verifiedDomains": undefined,
}
`;
